service/consul/environment.go

```diff
@@ -209,13 +208,69 @@ func NewEnvironment(l log.Logger, registrationScheme string, co Options, eo ...s
 		return nil, err
 	}
 
-	return environment{
+	newServiceEnvironment := environment{
 		service.NewEnvironment(
 			append(
 				eo,
 				service.WithRegistrars(r),
 				service.WithInstancers(i),
 				service.WithCloser(closer),
-			)...,
-		), NewClient(consulClient)}, nil
+			)...), NewClient(consulClient)}
+
+	if co.DatacenterWatchInterval > 0 {
+		go WatchInstancers(l, co, newServiceEnvironment)
```

Comment:
Personally, I've been moving toward Start() functions to start any long running goroutines, and Stop() functions to be able to properly stop them.  This helps with following the general advice to know when a goroutine exits:
https://github.com/golang/go/wiki/CodeReviewComments#goroutine-lifetimes

I'm not sure if that's needed here or not.

Reply:
I agree.  That's a nice pattern that lots of other libraries follow.  This code was written before a lot of those de facto standards became a thing.

It would be nice to backport this code to use the Start/Stop idiom, but only if the level of effort isn't too bad.